akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/AccountExampleWithCommandHandlersInState.scala

```diff
@@ -134,8 +136,8 @@ object AccountExampleWithCommandHandlersInState {
         throw new IllegalStateException(s"unexpected event [$event] in state [ClosedAccount]")
     }
 
-    def behavior(accountNumber: String): Behavior[AccountCommand[AccountCommandReply]] = {
-      EventSourcedBehavior.withEnforcedReplies[AccountCommand[AccountCommandReply], AccountEvent, Account](
+    def behavior[R <: AccountCommandReply](accountNumber: String): Behavior[AccountCommand[R]] = {
```

Comment:
What will be the type afterwards?

Reply:
Here a simpler variant:

```scala
def behavior(accountNumber: String): Behavior[AccountCommand[_]] = {
      EventSourcedBehavior.withEnforcedReplies[AccountCommand[_], AccountEvent, Account](
```

Reply:
That's then the only change necessary.

Reply:
Alternatively, I think the real fix should be in `ExpectingReply` converting the type parameter into an abstract type.

Reply:
What do you mean "what will be the type afterwards"?

Your version works for the test case, but by not changing ExpectingReply to be contravariant you're still limited.

We should try the abstract type approach, but we still need a type parameter for the Java DSL.

Reply:
A quick try with abstract types doesn't seem to work because of type inference in pattern matches is working good enough.

> Your version works for the test case, but by not changing ExpectingReply to be contravariant you're still limited.

Yes, that might be true.

> What do you mean "what will be the type afterwards"?

I meant the type after calling it because you will never have a concrete `R` available at the call-site so it will be basically equivalent to the version with the wildcard.

Reply:
You could have a concrete `OperationResult`, which gives you a `Behavior[AccountCommand[OperationResult]]` aka the behaviour for an actor that can receive `AccountCommand`s that return `OperationResult`s.

Reply:
But that's not what the code does, so `AccountCommand[_]` is the right type and also the one that is inferred with your code.

Reply:
You're right that right now my code doesn't require it (btw it infers `AccountCommand[Nothing]` not `AccountCommand[_]`).  But I believe it should be possible to legally (aka contravariantly, without casts) vary the reply type of commands.

Reply:
> But I believe it should be possible to legally (aka contravariantly, without casts) vary the reply type of commands.

I think it is. The actual usage will be de-facto contravariant. It's only the type that is not used anywhere else.

Reply:
WDYM by "it is" and "de-facto contravariant"?  Having `ExpectingReply` be invariant forces `AccountCommand` to be invariant which means it's not possible for `def behavior` to return anything other than `Behavior[AccountCommand[AccountCommandReply]]`.

Reply:
I meant that the reply type is not relevant in the behavior type, i.e. you should either try to hide it or use `Behavior[AccountCommand[_]]`. It is only used internally in an actor implementation. 

The `ExpectingReply` type is otherwise only used in the `reply` and `thenReply` for which it doesn't matter if it is declared contravariant or invariant. If you have a message of type `<: ExpectingReply[T]` you can call `reply` either with a `T` or some other type `U <: T`. That's what I meant with "it is" and "de-facto contravariant".

I guess it might be more of a problem if you want to abstract the call to `reply`.

Reply:
> I meant that the reply type is not relevant in the behavior type, i.e. you should either try to hide it or use `Behavior[AccountCommand[_]]`. It is only used internally in an actor implementation.

That sounds very much like "the message type is not relevant in the actor type, you should accept that actors can be sent any message".  Preserving the information and using the correct variance allows for greater and safer reuse, like the test shows.